### PR TITLE
Feat SES event-based email sendouts

### PR DESCRIPTION
**Pull request by tchalvak**
# Purpose of PR:

> --- Sending out emails via AWS SES via the conduit of EventBridge
- Email generated and sent (locally, but hooked into the AWS account eventbridge for testing purposes).
- It goes from: php code -> NMail php class -> send method -> aws-php-sdk -> eventbridge -> lambda -> SES
- And then in to the magic of the email system.

Fixes: https://github.com/BitLucid/ninjawars/issues/1230

End Result in my gmail:
image

## Before
We were using a system based on exim4 on the server, and also relying on a phpmailer library in php that became deprecated and then removed.

## After

Now we hook right in to SES via events.  We could probably eventually use SES templates and other such things.

## _Attached Screenshot of my change:_
image

image

## _Things that make review take longer:_

_(remove lines that do not apply to this PR)_

-   [x] Changing more than 20 files (much harder to review)
-   [x] Changing more than 5 files (a bit harder to review)
-   [x] Changes to critical code (login, dashboard, etc)
-   [ ] No comments on changed files (unknown)
-   Tests do not pass (will get pushed back)

## _Things that make review faster and easier:_

_(check box with an x where it applies)_

-   [x] I attached a screenshot of the changed part of the app working
-   [x] I added tests
-   [ ] This feature is requested specifically by a user
-   [x] This will fix a bug

## _Preview results in my branch at the url:_

-   Add aws access to send through SES?  You probably don't want to do that, so just review the tests and check that the new ones make sense, and if you're ambitious, run the unit tests yourself locally via: 
-  :/src/ninjawars$ ./vendor/bin/phpunit-watcher watch /src/ninjawars/deploy/lib/events/PutEventTest.php
-  You can run a signup at https://localhost:8765/signup , but of course that'll stop at the attempt to send an event to aws which will fail if you don't have it hooked in to an aws account.

### CODE REVIEW

**Comment by tchalvak** [comment]
@Beagle Feel like giving a PR review?  : D

**Comment by tchalvak** [comment]
Ah, need to run with some of this functionality, but do check through @Beagle  and give review and I’ll follow up.

File: deploy/lib/events/PutEvent.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+{
+    return new EventBridgeClient($config);
+}
+
+function sanitizeAndFormatEmail(array|string $email_complex): string
+{
+    list($email, $display) = is_array($email_complex) ?
+        [array_key_first($email_complex), reset($email_complex)] :
+        [$email_complex, null];
+    $sanitized_email = filter_var($email, FILTER_SANITIZE_EMAIL);
```
Comment (Beagle) [inline deploy/lib/events/PutEvent.php:43]: Double check this against valid email addresses that have special characters. As an example, `'` is a valid email character so you should be able to send to `seamus.o'malley@alfredjvazquez.us`

File: deploy/lib/events/PutEvent.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    if (false === $sanitized_email) {
+        throw new \Exception('Invalid email address: ' . $email);
+    }
+    return $display ? '"' . $display . '" <' . $sanitized_email . '>' : $sanitized_email;
+}
+
+/**
+ * Email config validation, which should perhaps eventually be moved away from event lib
```
Comment (Beagle) [inline deploy/lib/events/PutEvent.php:51]: What is config? Saying "validateEmailIncomingConfig provides email config validation" is not a useful comment. Perhaps: "This validates email header information includes from, to, subject, etc"

File: deploy/lib/events/PutEvent.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        "/\.|@/",
+        $to
+    )) {
+        return 'Invalid to email address: ' . $to;
+    }
+
+    return null;
+}
+
+function generateEmailEvent(array $config): array
```
Comment (Beagle) [inline deploy/lib/events/PutEvent.php:78]: "Returns a dictionary with the keys and data required to submit an event to the event bus"

File: deploy/lib/extensions/Nmail.class.php [hunk 4]
```diff
@@ -79,47 +96,23 @@ public function setReplyTo($email_or_array)
      *
      * @return boolean whether the mail function accepted the inputs.
      */
-    public function send()
+    public function send($debug_override = false)
```
Comment (Beagle) [inline deploy/lib/extensions/Nmail.class.php:99]: document the intention of debug_override

**Comment by Beagle** [comment]
Super nice. A great bridgehead into getting an EDA in place. Nice job!

**Pull request merged**